### PR TITLE
logind: create symlink from username to userid for runtime directory

### PR DESCRIPTION
**Pull request by eworm-de**
We create a directory /run/user/<userid>. Some programs can not handle
userid, though: I would like to place ssh control socket into this
directory, but ssh can expand '%u' to username only.
So let us create a symbolic link from username to userid.

### CODE REVIEW

File: src/login/logind-user.c [hunk 1]
```diff
@@ -383,6 +383,28 @@ static int user_mkdir_runtime_path(User *u) {
...
+        char *p;
+        int r;
+
+        assert(u);
+        assert(u->runtime_path);
+
+        if (asprintf(&p, "/run/user/%s", u->name) < 0)
+                 return log_oom();
+
+        if (is_symlink(p))
```
Comment (zonque) [inline src/login/logind-user.c:396]: `v` is leaked here. Best to use `_cleanup_free_ *p = NULL` in the declaration,

File: src/login/logind-user.c [hunk 2]
```diff
@@ -566,6 +593,23 @@ static int user_remove_runtime_path(User *u) {
         return r;
 }
 
+static int user_remove_runtime_symlink(User *u) {
+        char *p;
```
Comment (zonque) [inline src/login/logind-user.c]: `_cleanup_free_`

**Comment by eworm-de** [comment]
Updated to declare with `_cleanup_free_`.

Do we want that in `user_mkdir_runtime_path()` as well?

File: src/login/logind-user.c [hunk 3]
```diff
@@ -383,6 +383,28 @@ static int user_mkdir_runtime_path(User *u) {
...
+                 return log_oom();
+
+        if (is_symlink(p))
+                return 0;
+
+        r = symlink(u->runtime_path, p);
+        if (r < 0)
+                log_warning_errno(r, "Failed to symlink '%s'", p);
+
+        free(p);
```
Comment (zonque) [inline src/login/logind-user.c]: `_cleanup_free_` does that for you. As it's stands, it's a double free.

File: src/login/logind-user.c [hunk 4]
```diff
@@ -566,6 +593,23 @@ static int user_remove_runtime_path(User *u) {
...
+        assert(u);
+
+        if (asprintf(&p, "/run/user/%s", u->name) < 0)
+                 return log_oom();
+
+        r = unlink(p);
+        if (r < 0)
+                log_warning_errno(r, "Failed to remove runtime symlink '%s'", p);
+
+        free(p);
```
Comment (zonque) [inline src/login/logind-user.c]: Dito.

**Comment by eworm-de** [comment]
Ah, right... Updated again.

**Comment by poettering** [comment]
The only way to access /run/user/xyz is via $XDG_RUNTIME_DIR, and there's a specification for that. Never access this any other way.

http://standards.freedesktop.org/basedir-spec/basedir-spec-latest.html

Where precisely XDG_RUNTIME_DIR points to is up to the implementation, we do not consider /run/user/ API and take the liberty to change this any time deemed necessary without considering this an API brea.

**Pull request closed without merging**